rclcpp/include/rclcpp/experimental/serialization.hpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__EXPERIMENTAL__SERIALIZATION_HPP_
+#define RCLCPP__EXPERIMENTAL__SERIALIZATION_HPP_
+
+#include <rmw/rmw.h>
+
+#include <memory>
+#include <string>
+
+#include "rcl/error_handling.h"
+
+namespace rclcpp
+{
+namespace experimental
+{
+
+/// Interface to (de)serialize a message
+class SerializationBase
+{
+public:
+  virtual ~SerializationBase() {}
+
+  virtual std::shared_ptr<rcl_serialized_message_t> serialize_message(const void * message) = 0;
```

Comment:
yeah, I understand your concern. The Serialization class hides the rmw functions. It's difficult to avoid the void* without knowing the actual content (cannot be templated here). For the current PR we skip changing this

Reply:
The message is not converted in this sense. The resulting shared pointer is a new object. (And I want to prevent objects without cleanup functions as much as possible, so I prioritized preventing memory loss)

Reply:
I thought I had mentioned this somewhere else already, but what about using the `SerializedMessage` for this?
```
virtual void serialize_message(const void * ros_message, SerializedMessage & serialized_message);

virtual void deserialize_message(const SerializedMessage & serialized_message, void * ros_message);
```